subprojects/core/src/main/java/org/gradle/cache/internal/DefaultCrossBuildInMemoryCacheFactory.java

```diff
@@ -232,4 +242,52 @@ protected V maybeGetRetainedValue(Class<?> key) {
             return leakyValues;
         }
     }
+
+    private static class BuildAgeEvictionCrossBuildInMemoryCache<K, V> implements CrossBuildInMemoryCache<K, V>, SessionLifecycleListener {
+        private final ManualEvictionInMemoryCache<K, V> delegate = new ManualEvictionInMemoryCache<>();
+        private final ConcurrentMap<K, Boolean> keysFromPreviousBuild = new ConcurrentHashMap<>();
+        private final ConcurrentMap<K, Boolean> keysFromCurrentBuild = new ConcurrentHashMap<>();
+
+        @Override
+        public V get(K key, Function<? super K, ? extends V> factory) {
+            keysFromCurrentBuild.put(key, Boolean.TRUE);
+            return delegate.get(key, factory);
+        }
```

Comment:
I understand why we don't need to override `Cache.get(K, Supplier)` here, but it seems a bit fragile... wonder if there's a better solution? Capturing the delegation in a `ForwardingCache` base class could be one potential way.